### PR TITLE
[diffusion] model: support Z-Image-Omni

### PR DESCRIPTION
**Pull request by 66RING**
<!-- Thank you for your contribution! Please follow these guidelines to enhance your pull request. If anything is unclear, submit your PR and reach out to maintainers for assistance. Join our Slack community at https://slack.sglang.io to discuss further. -->

## Motivation

Support Z-Image-Omni

<!-- Describe the purpose and goals of this pull request. -->

## Modifications

<!-- Detail the changes made in this pull request. -->


## Accuracy Tests

<!-- If this pull request affects model outputs (e.g., changes to the kernel or model forward code), provide accuracy test results. -->


### SGL:

```bash
sglang generate --model-path Z-Image-Omni-Base  \
    --prompt "Young Chinese woman in red Hanfu, intricate embroidery. Impeccable makeup, red floral forehead pattern. Elaborate high bun, golden phoenix headdress, red flowers, beads. Holds round folding fan with lady, trees, bird. Neon lightning-bolt lamp (⚡️), bright yellow glow, above extended left palm. Soft-lit outdoor night background, silhouetted tiered pagoda (西安大雁塔), blurred colorful distant lights." \
    --height 1024 \
    --width 1024 \
    --num-inference-steps 40\
    --seed 42 \
    --guidance-scale 5 \
    --output-file-name "sgld_e0.png" \


sglang generate --model-path Z-Image-Omni-Base  \
    --prompt "Change the women's clothes to white cheongsam, keep other content unchanged" \
    --image-path "./outputs/sgld_e0.png" \
    --height 1280 \
    --width 768 \
    --num-inference-steps 40\
    --seed 42 \
    --guidance-scale 4 \
    --output-file-name "sgld_e1.png" \
```

### Diffusers:

```python
import os
import torch
import PIL
from diffusers import ZImageOmniPipeline

print("Creating pipeline...")
pipe = ZImageOmniPipeline.from_pretrained(
    checkpoint_path,
    torch_dtype=torch.bfloat16,
    low_cpu_mem_usage=False,
)

pipe.to("cuda")

# [Optional] Attention Backend
# Diffusers uses SDPA by default. Switch to Flash Attention for better efficiency if supported:
# pipe.transformer.set_attention_backend("flash")    # Enable Flash-Attention-2
# pipe.transformer.set_attention_backend("_flash_3") # Enable Flash-Attention-3

# [Optional] Model Compilation
# Compiling the DiT model accelerates inference, but the first run will take longer to compile.
# pipe.transformer.compile()

# [Optional] CPU Offloading
# Enable CPU offloading for memory-constrained devices.
# pipe.enable_model_cpu_offload()

# T2I
prompt = "Young Chinese woman in red Hanfu, intricate embroidery. Impeccable makeup, red floral forehead pattern. Elaborate high bun, golden phoenix headdress, red flowers, beads. Holds round folding fan with lady, trees, bird. Neon lightning-bolt lamp (⚡️), bright yellow glow, above extended left palm. Soft-lit outdoor night background, silhouetted tiered pagoda (西安大雁塔), blurred colorful distant lights."

print("Generating image...")
image = pipe(
    # image=[image_2],
    prompt=prompt,
    height=1024,
    width=1024,
    num_inference_steps=40,
    guidance_scale=5.0,
    generator=torch.Generator("cuda").manual_seed(42),
).images[0]

image.save("example_0.png")
print("Done!")

# I2I
prompt = "Change the women's clothes to white cheongsam, keep other content unchanged"
ref_image_1 = PIL.Image.open("./example_0.png").convert("RGB")

print("Generating image...")
image = pipe(
    image=[ref_image_1],
    prompt=prompt,
    height=1280,
    width=768,
    num_inference_steps=40,
    guidance_scale=4.0,
    generator=torch.Generator("cuda").manual_seed(42),
).images[0]

image.save("example_1.png")
print("Done!")
```



### Diff

testted commit id: a2b555f5fa038b56832b166d4a3d213165fb16dd

| zimage-omni-diffusers                                                                                                                      | zimage-omni-sglang                                                                                                                                                                                                                         |
|--------------------------------------------------------------------------------------------------------------------------------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| <img width="1024" height="1024" alt="example_0" src="https://github.com/user-attachments/assets/74b1c655-7001-44d1-8194-e09d4f84e23d" /> | <img width="1024" height="1024" alt="sgld_e0" src="https://github.com/user-attachments/assets/421aa80f-c745-47a0-8b11-f4c024b8429b" /> |
|<img width="768" height="1280" alt="example_1" src="https://github.com/user-attachments/assets/3677875d-6179-4b71-9dad-2e73def74351" /> | <img width="768" height="1280" alt="sgld_e1" src="https://github.com/user-attachments/assets/44fcb0dd-dc6a-40a7-9db3-d7aa36e9703b" /> |

| sgl-zimage-base | sgl-new-zimage-base |
|-------------|-----------------|
| <img width="1280" height="720" alt="zimage_base" src="https://github.com/user-attachments/assets/5885c63f-754f-41d9-970c-f04ceafbfaf6" /> | <img width="1280" height="720" alt="new_zimage" src="https://github.com/user-attachments/assets/961b922f-73f9-416d-ac8a-f307b8e2ed04" /> |




## Benchmarking and Profiling

<!-- If this pull request impacts inference speed, provide benchmarking and profiling results. -->

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).
- [ ] Work with maintainers to merge your PR. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process)

### CODE REVIEW

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/text_encoding.py [hunk 1]
```diff
@@ -240,17 +264,38 @@ def encode_text(
...
-            processed_text_list: list[str] = []
+            processed_text_list: list[str | list[str]] = []
             for prompt_str in texts:
-                preprocessed = preprocess_func(prompt_str)
-                processed_text_list.append(preprocessed)
+                if num_condition_images == 0:
+                    preprocessed = preprocess_func(prompt_str)
+                    processed_text_list.append(preprocessed)
+                else:
+                    # TODO: refactor, hard code for z-image-omni only for now.
```
Comment (66RING) [inline python/sglang/multimodal_gen/runtime/pipelines_core/stages/text_encoding.py]: break change.
Z-Image-Omni breaks a single prompt into multiple fragments.
so that
[
  frag0 + vae token + siglip token,
  frag1 + vae token + siglip token,
  ...
]

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/text_encoding.py [hunk 2]
```diff
@@ -321,15 +389,15 @@ def encode_text(
...
-        result.add_check(
-            "prompt_embeds", batch.prompt_embeds, V.list_of_tensors_min_dims(2)
-        )
-        result.add_check(
-            "negative_prompt_embeds",
-            batch.negative_prompt_embeds,
-            lambda x: not batch.do_classifier_free_guidance
-            or V.list_of_tensors_with_min_dims(x, 2),
-        )
+        # result.add_check(
```
Comment (66RING) [inline python/sglang/multimodal_gen/runtime/pipelines_core/stages/text_encoding.py]: Disable for debug omni mode. Since it break single prompt into multiple fragments. List[List[xxx]]

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/latent_preparation.py [hunk 3]
```diff
@@ -58,7 +60,11 @@ def forward(
         batch_size = batch.batch_size
 
         # Get required parameters
-        dtype = batch.prompt_embeds[0].dtype
+        dtype = (
```
Comment (66RING) [inline python/sglang/multimodal_gen/runtime/pipelines_core/stages/latent_preparation.py]: Single batch assertion may fail in zimage-omni mode.

File: python/sglang/multimodal_gen/runtime/pipelines_core/composed_pipeline_base.py [hunk 4]
```diff
@@ -309,10 +313,10 @@ def load_modules(
                 logger.info(
                     "Using custom VAE path: %s instead of default path: %s",
                     component_model_path,
-                    os.path.join(self.model_path, load_module_name),
+                    os.path.join(self.model_path, module_name),
```
Comment (66RING) [inline python/sglang/multimodal_gen/runtime/pipelines_core/composed_pipeline_base.py:291]: like:

```json
{
    "siglip": [
        "transformers",
        "Siglip2VisionModel"
    ],
}
```

1. module_name is siglip
2. wanna load it with transformer AutoModel
    - `load_module_name = module_name, map to transformer`
3. path is model_path/siglip
    - X `model_path/load_module_name = load_module_name/transformer`
    - O `model_path/module_name = load_module_name/siglip`

File: python/sglang/multimodal_gen/runtime/pipelines/zimage_pipeline.py [hunk 5]
```diff
@@ -113,4 +116,346 @@ def create_pipeline_stages(self, server_args: ServerArgs):
         )
 
 
-EntryClass = ZImagePipeline
+from sglang.multimodal_gen.runtime.models.vision_utils import load_image, load_video
+
+
+# TODO:
+# standalone for debug for now
```
Comment (66RING) [inline python/sglang/multimodal_gen/runtime/pipelines/zimage_pipeline.py]: Make it work first for now. Think about reuse later.

File: python/sglang/multimodal_gen/runtime/models/dits/zimage.py [hunk 6]
```diff
@@ -558,33 +926,333 @@ def patchify_and_embed(
...
+        if omni_mode:
+            assert len(unified_noise_mask) == 0, "Single batch only for now."
+            noise_mask_tensor = unified_noise_mask[0][:, : unified.shape[1]].unsqueeze(
+                0
+            )
+
+        return unified, unified_freqs, noise_mask_tensor
+
+    # TODO: copy to configs/zimage.py later
+    def get_freqs_cis(
```
Comment (66RING) [inline python/sglang/multimodal_gen/runtime/models/dits/zimage.py:1090]: break changes.

zimage omni supports multi-image input. So the embedding should be online compute one by one.

File: python/sglang/multimodal_gen/runtime/pipelines/zimage_pipeline.py [hunk 7]
```diff
@@ -52,20 +56,311 @@ def prepare_mu(batch: Req, server_args: ServerArgs):
     return "mu", mu
 
 
+class _ImageProcessStage(PipelineStage):
```
Comment (66RING) [inline python/sglang/multimodal_gen/runtime/pipelines/zimage_pipeline.py]: no idea how to reuse extra stage.

File: python/sglang/multimodal_gen/configs/pipeline_configs/zimage.py [hunk 8]
```diff
@@ -140,3 +140,205 @@ def prepare_pos_cond_kwargs(self, batch, device, rotary_emb, dtype):
...
+            truncation=True,
+            return_tensors="pt",
+        )
+
+        # TODO: hack
+        # prompt_list_lengths is used to reconstruct flattened_prompt into batching
+        inputs.prompt_list_lengths = prompt_list_lengths
+        token_lens = inputs.attention_mask.sum(dim=-1).tolist()
+
+        self.token_lens.append(token_lens)
```
Comment (66RING) [inline python/sglang/multimodal_gen/configs/pipeline_configs/zimage.py:454]: hack. since encoder can only return List[tensor]. we need to concat all slices and record this `slice len` as metadata.

File: python/sglang/multimodal_gen/configs/pipeline_configs/zimage.py [hunk 9]
```diff
@@ -157,3 +157,205 @@ def prepare_pos_cond_kwargs(self, batch, device, rotary_emb, dtype):
...
+        default_factory=lambda: (zimage_omni_postprocess_text,)
+    )
+
+    # TODO: review
+    # ugly hack
+    # pos token_lens, neg token_lens, pos token_lens, neg token_lens
+    # token_lens[0] for pos token_lens
+    # token_lens[1] for neg token_lens
+    # maybe bug in serving case
+    token_lens = []
```
Comment (66RING) [inline python/sglang/multimodal_gen/configs/pipeline_configs/zimage.py:383]: very hacky way. any suggestion? cc @yhyang201

File: python/sglang/multimodal_gen/runtime/models/dits/zimage.py (hunk 6 above)
Comment (66RING) [reply python/sglang/multimodal_gen/runtime/models/dits/zimage.py:1090]: freqs is online compute here, since we need latent and get pos_id from [condition, latent].

any suggestion? cc @yhyang201

File: python/sglang/multimodal_gen/runtime/models/dits/zimage.py [hunk 10]
```diff
@@ -590,42 +971,416 @@ def patchify_and_embed(
...
+        else:
+            # Single embedding for all tokens
+            adaln_input = self.t_embedder(t).type_as(x[0])
+            t_noisy = t_clean = None
+
+        # TODO: overwrite freqs_cis for debug
+        # TODO: single batch only
+        # compute freqs before patchify
+        # TODO: ugly
+        if freqs_cis is None:
```
Comment (66RING) [inline python/sglang/multimodal_gen/runtime/models/dits/zimage.py]: dit forward maybe need a review?

cc @yhyang201

File: python/sglang/multimodal_gen/runtime/pipelines/zimage_pipeline.py (hunk 7 above)
Comment (66RING) [reply python/sglang/multimodal_gen/runtime/pipelines/zimage_pipeline.py]: cc @yhyang201

File: python/sglang/multimodal_gen/runtime/pipelines/zimage_pipeline.py [hunk 11]
```diff
@@ -113,4 +117,385 @@ def create_pipeline_stages(self, server_args: ServerArgs):
...
+        if width % vae_scale != 0:
+            raise ValueError(
+                f"Width must be divisible by {vae_scale} (got {width}). "
+                f"Please adjust the width to a multiple of {vae_scale}."
+            )
+
+        # TODO:
+        # hard code with image size
+        # dulplicate with input_validation_stage
+        batch.height = image_height
```
Comment (66RING) [inline python/sglang/multimodal_gen/runtime/pipelines/zimage_pipeline.py]: batch.height and batch.width is hard code in InputValidationStage. 

So there is no way to adaptive get h,w from input image now.

any suggestion? cc @yhyang201

File: python/sglang/multimodal_gen/runtime/pipelines/zimage_pipeline.py [hunk 12]
```diff
@@ -113,4 +117,385 @@ def create_pipeline_stages(self, server_args: ServerArgs):
...
+        # should be condition_images(a list)
+        # condition_images -> prepare_image_latents
+        batch.condition_image = condition_image
+        # prepare for
+        # resized_images -> prepare_siglip_embeds
+        batch.resized_images = resized_images
+        return batch
+
+
+class _PrepareImageLatentsStage(PipelineStage):
```
Comment (66RING) [inline python/sglang/multimodal_gen/runtime/pipelines/zimage_pipeline.py]: can I just code standalone stage in here? Since I have no idea how to reuse it.

any suggestion? cc @yhyang201

File: python/sglang/multimodal_gen/runtime/pipelines_core/composed_pipeline_base.py (hunk 4 above)
Comment (66RING) [reply python/sglang/multimodal_gen/runtime/pipelines_core/composed_pipeline_base.py:291]: any suggestion? cc @yhyang201
